app/v2/controllers/UpscanResponseParser.scala

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v2.controllers
+
+import cats.data.EitherT
+import play.api.Logging
+import play.api.libs.json.JsValue
+import play.api.mvc.BaseController
+import v2.models.errors.PresentationError
+import v2.models.responses.UpscanResponse
+
+import scala.concurrent.Future
+
+trait UpscanResponseParser {
+  self: BaseController with Logging =>
+
+  def parseAndLogUpscanResponse(responseBody: JsValue): EitherT[Future, PresentationError, UpscanResponse] = {
+    println("parseAndLogUpscanResponse...")
+    EitherT {
+      responseBody
+        .validate[UpscanResponse]
+        .map {
+          upscanResponse =>
+            logResponse(Some(upscanResponse))
+            Future.successful(Right(upscanResponse))
+        }
+        .getOrElse {
+          logResponse(None)
+          Future.successful(Left(PresentationError.badRequestError("Unexpected Upscan callback response")))
+        }
+    }
+  }
+
+  private def logResponse(upscanResponse: Option[UpscanResponse]) =
```

Comment:
I suppose this works, but it's a bit weird that you're handling code from two different branches, that'll do two different things, in this function -- you'll only get `None` from the `getOrElse` part of the function above, you'll only get `Some` from the `map` part. 

I think this should just accept a `UpscanResponse` and deal with success cases. The failure case (`None`) is a simple one-liner that should go in the appropriate place in the function above -- it's a little less complicated that way.

It also means you don't have to re-wrap the argument in `map` with a `Some`.

Reply:
yes, it's looks much simpler now.